internal/status/status.go

```diff
@@ -0,0 +1,318 @@
+	return &statusHelper{
+		deploymentAPI: deploymentAPI,
+		daemonsetAPI:  daemonsetAPI,
+	}
+}
+
+func (sh *statusHelper) getDaemonSetNotAvailableConditions(ctx context.Context, dsNamespace, dsName string) []metav1.Condition {
+	failedToGetDSReason := conditionFailedGettingNFDWorkerDaemonSet
+	dsDegradedReason := conditionNFDWorkerDaemonSetDegraded
+	dsProgressingReason := conditionNFDWorkerDaemonSetProgressing
+	if dsName == "nfd-topology-updater" {
+		failedToGetDSReason = conditionFailedGettingNFDTopologyDaemonSet
+		dsDegradedReason = conditionNFDTopologyDaemonSetDegraded
+		dsProgressingReason = conditionNFDTopologyDaemonSetProgressing
+	}
+	ds, err := sh.daemonsetAPI.GetDaemonSet(ctx, dsNamespace, dsName)
+	if err != nil {
+		return getDegradedConditions(failedToGetDSReason, err.Error())
+	}
+	conditionsStatus, message := getDaemonSetConditions(ds)
+	if conditionsStatus == conditionStatusDegraded {
+		return getDegradedConditions(dsDegradedReason, message)
+	} else if conditionsStatus == conditionStatusProgressing {
+		return getProgressingConditions(dsProgressingReason, message)
+	}
+	return nil
+}
+
+func (sh *statusHelper) getDeploymentNotAvailableConditions(ctx context.Context, deploymentNamespace, deploymentName string) []metav1.Condition {
+	failedToGetDeploymentReason := conditionFailedGettingNFDMasterDeployment
+	deploymentDegradedReason := conditionNFDWorkerMasterDeploymentDegraded
+	deploymentProgressingReason := conditionNFDMasterDeploymentProgressing
+	if deploymentName == "nfd-gc" {
+		failedToGetDeploymentReason = conditionFailedGettingNFDGCDeployment
+		deploymentDegradedReason = conditionNFDGCDeploymentDegraded
+		deploymentProgressingReason = conditionNFDGCDeploymentProgressing
+	}
+	dep, err := sh.deploymentAPI.GetDeployment(ctx, deploymentNamespace, deploymentName)
+	if err != nil {
+		return getDegradedConditions(failedToGetDeploymentReason, err.Error())
+	}
+	conditionsStatus, message := getDeploymentConditions(dep)
+	if conditionsStatus == conditionStatusDegraded {
+		return getDegradedConditions(deploymentDegradedReason, message)
+	} else if conditionsStatus == conditionStatusProgressing {
+		return getProgressingConditions(deploymentProgressingReason, message)
+	}
+	return nil
+}
+
+func getDaemonSetConditions(ds *appsv1.DaemonSet) (string, string) {
+	if ds.Status.DesiredNumberScheduled == 0 {
+		return conditionStatusDegraded, "number of desired nodes for scheduling is 0"
+	}
+	if ds.Status.CurrentNumberScheduled == 0 {
+		return conditionStatusDegraded, "0 nodes have pods scheduled"
+	}
+	if ds.Status.NumberReady == ds.Status.DesiredNumberScheduled {
+		return conditionStatusAvailable, ""
+	}
+	return conditionStatusProgressing, "ds is progressing"
+}
+
+func getDeploymentConditions(dep *appsv1.Deployment) (string, string) {
+	if dep.Status.AvailableReplicas == 0 {
+		return conditionStatusDegraded, "number of available pods is 0"
+	}
```

Comment:
Do we have support for more than one replica (in the operator)? Should we  add similar StatusProgressing condition as for the daemonset?

Reply:
Currently operator deploys with replica of 1 and it is not configurable by user.  
In regards to Progressing vs Degraded: the distinction is blurred, but in this PR i was adhering to the current implementation. It is not urgent, but i think that we can discuss if we really need both Degraded and Progressing.How we distinguish between the 2: in both cases we don't have enough pods running, and K8S will try to bring up additional pods, so we are both in Degraded state and in Progressing state...